src/passes/SignatureSubtyping.cpp

```diff
@@ -0,0 +1,198 @@
+// This differs from DeadArgumentElimination's refineArgumentTypes() etc. in
+// that DAE will modify the type of a function. It can only do that if the
+// function's type is not observable, which means it is not taken by reference.
+// On the other hand, this pass will modify the signature types themselves,
+// which means it can optimize functions whose reference is taken, and it does
+// so while considering all users of the type (across all functions sharing that
+// type, and all call_refs using it).
+//
+// TODO: optimize results too and not just params.
+//
+
+#include "ir/find_all.h"
+#include "ir/lubs.h"
+#include "ir/module-utils.h"
+#include "ir/type-updating.h"
+#include "pass.h"
+#include "wasm-type.h"
+#include "wasm.h"
+
+using namespace std;
+
+namespace wasm {
+
+namespace {
+
+struct SignatureSubtyping : public Pass {
+  // Maps each heap type to the possible refinement of the types in their
+  // signatures. We will fill this during analysis and then use it while doing
+  // an update of the types. If a type has no improvement that we can find, it
+  // will not appear in this map.
+  std::unordered_map<HeapType, Signature> newSignatures;
+
+  void run(PassRunner* runner, Module* module) override {
+    if (getTypeSystem() != TypeSystem::Nominal) {
+      Fatal() << "SignatureSubtyping requires nominal typing";
+    }
+
+    // First, find all the calls and call_refs.
+
+    struct CallInfo {
+      std::vector<Call*> calls;
+      std::vector<CallRef*> callRefs;
+    };
+
+    ModuleUtils::ParallelFunctionAnalysis<CallInfo> analysis(
+      *module, [&](Function* func, CallInfo& info) {
+        if (func->imported()) {
+          return;
+        }
+        info.calls = std::move(FindAll<Call>(func->body).list);
+        info.callRefs = std::move(FindAll<CallRef>(func->body).list);
+      });
+
+    // A map of types to the calls and call_refs that use that type.
+    std::unordered_map<HeapType, CallInfo> allCallsTo;
+
+    // Combine all the information we gathered into that map.
+    for (auto& [func, info] : analysis.map) {
+      // For direct calls, add each call to the type of the function being
+      // called.
+      for (auto* call : info.calls) {
+        allCallsTo[module->getFunction(call->target)->type].calls.push_back(
+          call);
+      }
+
+      // For indirect calls, add each call_ref to the type the call_ref uses.
+      for (auto* callRef : info.callRefs) {
+        auto calledType = callRef->target->type;
+        if (calledType != Type::unreachable) {
+          allCallsTo[calledType.getHeapType()].callRefs.push_back(callRef);
+        }
+      }
+    }
+
+    // Compute optimal LUBs.
+    std::unordered_set<HeapType> seen;
+    for (auto& func : module->functions) {
+      auto type = func->type;
+      if (seen.count(type)) {
+        continue;
+      }
+      seen.insert(type);
+
+      auto sig = type.getSignature();
+
+      auto params = sig.params;
+      auto numParams = params.size();
```

Comment:
It looks like `params` isn't used after this, so it probably doesn't need to be named.
```suggestion
      auto numParams = sig.params.size();
```